### ISSUE TITLE
Pipelines : suggest reworking behaviour when reader can't progress and writer is throttled

### ISSUE THREAD

**Issue by mgravell** [dotnet/corefx#30689]
Situation: the reader can't sensibly consume the existing data, presumably because it needs more data for a semantic frame, but the `PauseWriterThreshold` has been reached, so the writer is inactive.

What happens today:

>  System.InvalidOperationException : Advancing examined to the end would cause pipe to deadlock because FlushAsync is waiting.

This is very non-obvious to the casual reader - in particular, it is misleading about the problem:

- the writer is correctly using the API - they have called `FlushAsync()` and are behaving properly
- the reader is correctly using the API - they have signalled what data they have inspected, and that they cannot progress

The *actual problem* currently is: the `PauseWriterThreshold` is currently too small for the application's requirements. IMO the exception should *focus on this*, rather than blaming the reader.

---

A casual "solution" might be:

> oh, the reading code just needs to consume some of the data in it's own backlog

but to my mind that *completely defeats the point of pipelines* - the emphasis being that the pipe *is* the backlog if you can't do something partial; the workaround - especially for binary - would ultimately be either to duplicate pipelines functionality, or start allocating / leasing separate buffers. So I don't think that's a good solution

---

My *preferred* solution here would be: if the reader *has touched the right-hand edge* (i.e. they've signalled that they've examined the range and can't use it) - specifically, something like:

    reader.AdvanceTo(..., buffer.End);

(the `buffer.End` being the important bit here), then `ReadAsync` is allowed to awaken the writer, and the writer is allowed to push more data; when they next `FlushAsync`, the reader is allowed to re-test for more data. This is still using the "writer is throttled" model, but it automatically resolves this situation, while allowing pipelines to do all the great stuff re backlog management and avoiding additional allocations.

**Comment by mgravell** [comment]
(/cc @davidfowl)

**Comment by mgravell** [comment]
Additional observation inspired by a @Drawaes conversation (ok, this might be entirely his point, not mine): the current "workaround" of setting a huge `PauseWriterThreshold` has a nasty side-effect of allowing the writer to over-fill the point even when the reader **isn't** hitting this problem. His example was a parser on a file-stream - by having a large `PauseWriterThreshold`, the writer **unnecessarily** basically pushed the entire file into memory.

**Comment by Drawaes** [comment]
Agreed. To avoid any "potentially" large record that "may" come up causing a crash you have to always over commit on the buffering so I agree with all above that the "examined" not the "consumed" should be where the high water mark is placed.

**Comment by joshfree** [comment]
@pakrym

**Comment by mgravell** [comment]
To rephrase this:

The point of the threshold is to slow the writer when the reader isn't keeping up, to prevent it over-filling a pipe without reason.

But if the reader has "examined" the end, they *are* keeping up: they're basically saying "keep it coming, I need more! MOAR!". That's why it is a slightly different situation, and one that IMO should be allowed.

**Comment by davidfowl** [comment]
I think we should consider this for 2.2

**Comment by mgravell** [comment]
Note: it probably makes sense for there to be two settings instead of just the one:

- the threshold for when the reader isn't keeping up
- the absolute maximum pipe size (at which an error legitimately occurs)

The current setting is *essentially* acting like the second of these, although in many ways it claims to be the first. Since people hitting this issue will have *changed that value to fix it*, I suggest that it would make sense to keep this setting meaning "the absolute maximum", but changing the default to something much bigger, like 1GB; the *new* setting the would take the current default of 32kB.

Possibly. Or not. Just thinking aloud here.

**Comment by davidfowl** [comment]
So after thinking about this and discussing it with a few people it seems as though there are a couple of concerns here:
- Understanding back pressure is complicated as is and adding yet another limit introduces more concepts. What you're suggesting is a soft limit and a hard limit. What we have today is a hard limit that throws if more than that amount if buffered. (Also the exception message sucks)
- Hitting this soft limit will pause the writer on every write past the limit if the reader doesn't consume down the to resume threshold. This is the most inefficient way to read/write since you'll be paused on each write past the max.
- It makes it hard to predict how much memory the system will use if the limit isn't really the limit.

There are a couple of other mitigations that you could use:

- If you own the pipe creation, it's better to just bump the limit up to what the max message size would be. This is the best option from a performance POV but you need to know the maximum message size your protocol supports.
- If the pipe settings are out of your control the create a new pipe that copies from the original pipe so that you control the limit and can set it to the max message size. This sucks because you incur needless copies.

Some other options that were discussed:
- An overload of ReadAsync that lets you override whatever maximum buffer size might exist internally. This would give control per call to read.

**Comment by mgravell** [comment]
> An overload of ReadAsync that lets you override whatever maximum buffer size might exist internally. This would give control per call to read.

I doubt that `ReadAsync` could ever know, since it would likely be making this determination based on the contents of the incoming frame(s); an overload of `AdvanceTo` seems a more likely place if you wanted to allow an "I want to rock the boat by {this much}" flag. Although I guess you *could* store that and tell it at the *next* `ReadAsync`.

When looking at it *this* way, it is actually broadly related to [Ability to wait for a specific amount of data](https://github.com/dotnet/corefx/issues/271850) - but with the observation that sometimes "specific amount of data" might be larger than the regular throttle limit.

**Comment by mgravell** [comment]
@davidfowl  also, the "Ability to wait for a specific amount of data" **totally fixes** the:

> This is the most inefficient way to read/write since you'll be paused on each write past the max.

issue, as it will only need to do it *once*.

**Comment by mgravell** [comment]
And it removes the need for a separate configuration option that adds confusion. Ooh, this would be a great two-for-one :)

**Comment by Drawaes** [comment]
Yeah and it should reduce the thread wake ups needed (which is why I wanted it initially) so I would be very happy with that. Should I start a proper API proposal for it?

**Comment by mgravell** [comment]
> Should I start a proper API proposal for it?

proposal: add `AdvanceTo(SequencePosition consumed, SequencePosition examined, long bytesNeeded)` overload; implements [magic.gif]

**Comment by pakrym** [comment]
What about adding a way to set limits dynamically to the `Pipe`? This way you can implement the soft limit outside the pipe by expanding limits when a deadlock is about to occur.

**Comment by Drawaes** [comment]
How is that less complex than two limits or anything else suggested above. Seems like that would be a complexity nightmare

**Comment by pakrym** [comment]
They are still simple to reason about - work as original limits just are changeable and the change is explicit.

And we need dynamic limits for other use cases so if we can fix two problems with one feature it would be nice.

**Comment by Drawaes** [comment]
Does the limit "reset" or is it for the lifetime of the connection and you need to reset?

**Comment by AArnott** [comment]
> Understanding back pressure is complicated as is and adding yet another limit introduces more concepts. What you're suggesting is a soft limit and a hard limit. What we have today is a hard limit that throws if more than that amount if buffered.

I would resolve that by *not* having two limits. Just make the currently hard limit a soft limit instead.

> the "Ability to wait for a specific amount of data" totally fixes ...

I'm not a big fan of this solution. In @davidfowl's blog post, he described a reader that looks for EOL characters. If the 32KB limit is perfectly sane for every file except this js-minified file that is 64KB, and this reader is looking for EOL characters, it couldn't predict how much data will be needed to find the EOL character. I realize this can mitigate the reader/writer thrashing issue, but I don't think we can require the reader to know how much more to read before waking it up (or before it starts to block the writer again). If it knows, then great. Otherwise, I think thrashing in that scenario may be acceptable. It's obviously a very unusual situation (if it wasn't, you should raise the pause threshold).

**Comment by AArnott** [comment]
BTW here is the unit test I wrote to explore this scenario (hoping that the limit would be a soft limit) which led me to see the actual behavior and find this github issue:

```cs
[Fact]
public async Task MyTest()
{
    Pipe pipe = new Pipe(new PipeOptions(pauseWriterThreshold: 5, resumeWriterThreshold: 2));
    await Task.WhenAll(
        Task.Run(async () =>
        {
            for (int i = 0; i < 10; i++)
            {
                this.Logger.WriteLine($"Writing byte {i + 1}");
                await pipe.Writer.WriteAsync(new byte[1], this.TimeoutToken);
                this.Logger.WriteLine($"Flushing byte {i + 1}");
                await pipe.Writer.FlushAsync(this.TimeoutToken);
            }

            pipe.Writer.Complete();
        }),
        Task.Run(async () =>
        {
            while (true)
            {
                ReadResult readResult = await pipe.Reader.ReadAsync(this.TimeoutToken);
                this.Logger.WriteLine($"Reading returned {readResult.Buffer.Length} bytes.");
                if (readResult.Buffer.Length < 10)
                {
                    pipe.Reader.AdvanceTo(readResult.Buffer.Start, readResult.Buffer.End);
                }
                else
                {
                    pipe.Reader.AdvanceTo(readResult.Buffer.End, readResult.Buffer.End);
                }
            }
        }));
}
```

**Comment by davidfowl** [comment]
It’s extremely difficult to predict the memory usage of an application if the limits are just soft limits. This can lead to security issues in the worse case if remote client can basically force the server to buffer indefinitely.

Currently you need to have a max message size that is less than the pause writer threshold. That’s completely logical as the entire point of that property is to limit the memory. The 32K default was pretty arbitrary and was basically there just to prevent people from DOSing yourself by default.

**Comment by ciddan** [comment]
I ran into exactly this issue when trying out pipelines today, result in the following code which basically defeats the entire purpose of using Pipe in the first place:

```C#
// messageLength is from the first 4 bytes of the stream
if (buffer.Length < messageLength)
{
	buffer = buffer.Slice(constTotalHeaderLength);

	var messageBuffer = MemoryPool<byte>.Shared.Rent((int) messageLength);
	var bufferedSize = 0L;
	var messageConsumed = false;

	while (true)
	{
		if (bufferedSize + buffer.Length >= messageLength)
		{
			buffer = buffer.Slice(0, messageLength - bufferedSize);
			buffer.CopyTo(messageBuffer.Memory.Span.Slice((int) bufferedSize));
			messageConsumed = true;
		}
		else
		{
			buffer.CopyTo(messageBuffer.Memory.Span.Slice((int) bufferedSize));
			bufferedSize += buffer.Length;
		}

		aPipeReader.AdvanceTo(buffer.GetPosition(buffer.Length));
		if (messageConsumed)
		{
			break;
		}

		result = await aPipeReader.ReadAsync(aCancellationToken);
		buffer = result.Buffer;
	}

	var decodedMessage = myEncoderDecoder.Decode<TIn>(
		messageBuffer.Memory.Slice(0, (int) messageLength).ToArray());
	aMessageDecodedCallback(requestId, decodedMessage);

	messageBuffer.Dispose();
	continue;
}
```
The discrepancy between `buffer.Length` and `buffer.End` also threw me for a loop for a while. buffer.Length could be 16k while buffer.End was 2k.

**Comment by davidfowl** [comment]
> I ran into exactly this issue when trying out pipelines today, resulting in the following code which basically defeats the entire purpose of using Pipe in the first place:

@ciddan why didn't you just increase the threshold?

> PS: The discrepancy between buffer.Length and buffer.End also threw me for a loop for a while. buffer.Length could be 16k while buffer.End was 2k.

I'm confused here. What do you mean "while buffer.End" was 2k?

**Comment by ciddan** [comment]
@davidfowl 
1) Because the threshold isn't deterministic. I can't guarantee the size of message that will be sent to me (API outside of my control).
2) I was trying to "do the right thing" and only use `buffer.End` for Slicing and Advancing, but Slicing using buffer.End would sometimes give me more bytes than the 2048 indicated by buffer.End - up to the full 32k default limit.

Doing this:
```C#
buffer.Slice(buffer.Start, buffer.End).CopyTo(messageBuffer.Memory.Span);
```
could result in way more than 2048 bytes being copied into `messageBuffer`.

**Comment by davidfowl** [comment]
> Because the threshold isn't deterministic. I can't guarantee the size of message that will be sent to me (API outside of my control).

Don't you have that same problem with the code you wrote? If that code represents a server, you'll just rent a single array from the default memory pool of the length specified by the client. How is this logic any better than increasing the limit?

> I was trying to "do the right thing" and only use buffer.End for Slicing and Advancing, but Slicing using buffer.End would sometimes give me more bytes than the 2048 indicated by buffer.End - up to the full 32k default limit.

I'm confused by this. Of course it would give you more than 2048, why did you think it wouldn't? This diagram was supposed to show that:

![image](https://user-images.githubusercontent.com/95136/42292592-74a4028e-7f88-11e8-85f7-a6b2f925769d.png)

Lets say you're using the default settings 2 KiB segment size and 32 KiB limit then (buffer.End - buffer.Start) could represent the entire 32K buffer across the 16 segments of memory.

**Comment by ciddan** [comment]
Let's just say all the nuances of the API aren't immediately apparent after reading your article and the simplified use case of the example repo. Remember, we haven't all been working with Pipeline for an extended period of time. To me, the intuitive thing after reading the example code was to rely on buffer.End when Slicing. Especially when buffer.End.GetInteger() would give me 2048 back, but when Slicing with it, it could give me up to the 32k limit.

**Comment by ciddan** [comment]
> Good to know. You should never use GetObject or GetInteger ever. We debated for a long time about removing those APIs but we chose to hide them from intellisense instead. Those APIs exists for the internal implementation details of the thing that created the ReadOnlySequence. That's the only thing that can interpret those values correctly.

Haha :) They aren't hidden from IntelliSense in Rider. I'll post a bug on their tracker about that. I assumed it was a standard API since it showed up in Rider when I "dotted" `buffer`. Good to know.

> Do you want to allow any size message or do you have a maximum message size in mind?
I'm in that situation now, yes. The issue with my use case, after thinking about it again, likely isn't with Pipe, but with the C# Protobuf implementation. There's some discussion ongoing on their repo about adding support for byte*, Span<byte> or Memory<byte> - which may solve my issue. At the moment I have to collect the full message (copying in to a buffer), max size unknown (at compile time), from a TCP socket and pass it in to the Protobuf lib for parsing. On the originating C++ side they have access to ZeroCopyStream protobuf API's which aren't present on the C# side.

**Comment by davidfowl** [comment]
Do you own both the client and the server in this case? The limits are there to basically prevent you from creating a situation where you let an untrusted writer, cause the reader to buffer too much (resulting in a DOS). If it's fully trusted, you can consider removing the limits completely. Then the only question is about performance of small vs large messages.

**Comment by ciddan** [comment]
The server isn't owned by my team, but it is owned by the organization - so it is trusted in that sense. The solution to this, I suppose, is to ask the other team to implement hard pagination limits on the data, so that we can impose a guaranteed max response size from their API, or implementing our own caching/paginating middle layer in between.

**Comment by davidfowl** [comment]
Or just disable the limits until further notice and monitor the memory.

**Comment by tmds** [comment]
I had encountered this when playing with corefxlab pipelines: https://github.com/dotnet/corefxlab/issues/1355

**Comment by davidfowl** [comment]
I do wonder if the default limit is too small now. Maybe we should bump it to 1MB or 4MB

**Comment by Drawaes** [comment]
I suspect if you do that 90% of these reports will just disappear.

**Comment by mgravell** [comment]
> Or just disable the limits until further notice and monitor the memory.

The one thing I want to reiterate about that as a workaround is that it allows the writer to **unnecessarily** swamp the pipe. If the reader needs more data, I'm fine with that - it is "necessary" at that point; but that will often be the exception not the rule, and there's no need for the pipe to be stuffed full the rest of the time - which upping or removing the quota may achieve.

**Comment by mgravell** [comment]
@ciddan are you using the Google lib? Or protobuf-net? Or another implementation?

**Comment by Drawaes** [comment]
This is my main issue with the argument. I want a "normally only readahead x amount" but I might need more than that for one message.

**Comment by ciddan** [comment]
@mgravell The google lib.

**Comment by davidfowl** [comment]
> This is my main issue with the argument. I want a "normally only readahead x amount" but I might need more than that for one message.

My problem is, there needs to be a hard limit regardless and you still need to configure the soft limit to *something*. If you don't have a good feel for that, then the problem here isn't resolved. The underlying problem here is that we have a low default the everyone is hitting and I think that's the core of the problem here. If people don't know what to set the soft limit to, how is it any better? Other than giving a false sense of security (for those that need it) .

**Comment by davidfowl** [comment]
I'm not sure everyone does but I before adding a soft limit the things we're going to fix are:
- Increasing the default 32K limit to something like a 1MB instead https://github.com/dotnet/corefx/issues/31101
- Fix the error message so that it's clear what happened https://github.com/dotnet/corefx/issues/30894

**Comment by Drawaes** [comment]
Why don't you add a "new" limit.

readAheadThreshold

pauseWriterThreshold == how much "unconsumed" data you can have
readAheadThreshold == how much "unexamined" data you can have

Then people can set the readAheadThreshold to a reasonable number (32k or whatever) and the other one to the large limit of the upperbound they may expect to process?

Simples :)

**Comment by benaadams** [comment]
Remove the exception for `pauseWriterThreshold`; have it just be when backpressure starts being applied.

Is a valid change:

> [Breaking Change Rules](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#exceptions)
> Allowed:
> Removing an exception that was being thrown when the API allows more robust behavior or enables new scenarios

Add a backstop threshold `maxUnconsumedThreshold` when `GetMemory` starts to throw

**Comment by tmds** [comment]
My take on this (based on https://github.com/dotnet/corefxlab/issues/1355).

Add an additional property on the Pipe `MaxLength`. If the writer goes beyond it, the pipe explodes.

When the reader advances to the end (`reader.AdvanceTo(..., buffer.End);`), then the writer is unblocked.

> Hitting this soft limit will pause the writer on every write past the limit if the reader doesn't consume down the to resume threshold. This is the most inefficient way to read/write since you'll be paused on each write past the max.

It's hard to assess this. If the producer is producing somewhat meaningful chunks, maybe this is not much of a problem. Or maybe, it is relatively cheap to be ignored by the reader (compared to producing the data).

> It makes it hard to predict how much memory the system will use if the limit isn't really the limit.

If not in the pipe, it needs to store it elsewhere. The system stores the same data, just in different places.

> an overload of AdvanceTo seems a more likely place

Yes. This seems a good place to add a parameter that controls how much the writer can push into the pipe. This is a _maximum_ that updates the pausewriterthreshold.

This parameter may be different to the _mimimum_ the reader wants to get from its next call to ReadAsync (https://github.com/dotnet/corefx/issues/27185).

**Comment by GSPP** [comment]
Many applications have a small message size normally but must support a large message size occasionally. Memory usage through buffering should then be small in the common case but be able to expand to very large values to support the large case.

If avg. message size is 1KB and max message size is 100MB then we don't want the pipe to buffer 100MB if not needed.

There need to be two numbers:

1. Set the default buffer size for *efficiency* (here maybe 32KB or 1MB).
2. Set a limit for reliability and security reasons (100MB).

Marc's solution seems to properly address this.

**Comment by davidfowl** [comment]
> Set a limit for reliability and security (100MB).

That's way too big. If you're talking about a sever, how many connections would it take in order to eventually DOS the server? Not many...

I still hold that the problem is with the defaults, nobody wants to think about setting them so when there's an issue that requires you think about a number, you're back to square one. Not only do you have to think about the max buffer size, you also have to think about your average buffer size now just in case our defaults didn't account for your specific scenario.

On top of that, you won't even know that the defaults for the pause threshold are causing you problems because it won't throw, your performance might just be worse.

**Comment by MaxDZ8** [comment]
I might have lost track of how this is progressing but it seems to me thresholds are getting most of the discussion.

What was the problem with
> proposal: add `AdvanceTo(SequencePosition consumed, SequencePosition examined, long bytesNeeded)` overload; implements [magic.gif]
?

In this situation, we are looking for a way to go beyond FULL. To this degree, I would say `AdvanceTo(SequencePosition consumed, SequencePosition examined, long bytesNeeded)` provides a better way to finely adjust the over-full-capability. I definitely prefer it over working with limits and it seems to me it's more sustainable long term as the extended limit could be relinquished on next consume automatically. No, I don't want to even think with what this implies as pinned hardware buffers are in the mix, if you use them.

**Comment by ZeoJF** [comment]
Ran into this problem today.
Reading through the above it occurs to me that the current implementation is trying to manage buffering with the writer to the detriment of the reader; the entity that knows the buffering requirements.

David is concerned about DOS and write performance, I like that, so let write buffering mean I will keep X bytes of unread data available.
Change ReadAsync so that it can request a minimum chunk size, kind of like writer.getmemory(),  the reader knows what that requirement is and is in effect doing its buffering in the pipe.  You'll have to do something if the last chunk doesn't meet the minimum but it won't be block.
Remove the confusing bits of AdvanceTo.  You only need consumed which gives the reader its mechanism to say, "throw away this much of the buffer, I'm done with it."

Basically, if buffering needs to happen, the reader is the only one who knows it and to suggest that I build a buffer and read into it in small chunks subverts the idea that I can use Pipes as a buffer mechanism.

So, there's room for the writer to stay ahead but not go overboard (DOS) and the reader controls buffer mechanics. (which they do anyway - even in the streams world view)  As to predicting memory usage, that's a problem for the reader, not the pipe(buffer) or writer.

I'm new to Pipes (Well these pipes anyway) so I could be off here.

**Comment by ZeoJF** [comment]
I thought about this a bit overnight.  If you keep ReadAsync() (unbounded, without a requestSize) you'll need to limit the result to a maximum as unbounded reads will keep pushing on the write ahead space and negate the DOS protection.

Also, I don't want to know about your segmented memory model or memory pools or the implementation of flow-control.  (yeah, all the cool stuff really)  To show you what I mean in a language I think we all understand:

All made up just to show a vision.  I see I already forgot to Release on the reads, but you should get the drift.  I should be passing through a cancellationToken but left that out and I'm not sure the value of asking my side of a pipe if the other side closed because it was cancelled so I left that out too.

    // acts kinda like Memory<byte>
    // but allows hidden magic implementations
    public interface IPipeWriteBuffer
    {
        // this or
        Memory<byte> AsMemory();
        // this or
        Memory<byte> Slice(int start, int length); // I like this for consistency
        // maybe this makes more sense
        void CopyFrom(Memory<byte> bytes);

        // left for the reader
    }

    // disposing, a well known construct, is how the pipe closes this end
    public interface IPipeWriter : IDisposable
    {
        // returns default if Pipe is closed
        IPipeWriteBuffer GetWriteBuffer(int sizeHint);
        // returns false if write didn't happen.
        ValueTask<bool> FlushAsync(int bytesWritten);
    }

    public static partial class PipeWriterExtensions
    {
        public static async ValueTask<bool> SendAsync(
            this IPipeWriter writer,
            IEnumerable<(int hint, Func<IPipeWriteBuffer, int> fill)> chunks)
        {
            var success = true;

            foreach (var (hint, fill) in chunks)
                success = await writer.SendAsync(hint, fill).ConfigureAwait(false);

            return success;
        }

        public static async ValueTask<bool> SendAsync(
            this IPipeWriter writer,
            int sizeHint,
            Func<IPipeWriteBuffer, int> fill)
        {
            var success = true;

            IPipeWriteBuffer buffer;
            if (0 < sizeHint && default != (buffer = writer.GetWriteBuffer(sizeHint)))
                success = await writer.FlushAsync(fill(buffer)).ConfigureAwait(false);

            return success;
        }
    }

    // acts just like ReadOnlyMemory<byte>
    // but allows hidden magic implementations
    public interface IPipeReadBuffer
    {
        // left for the reader
    }

    // disposing, a well known construct, is how the pipe closes this end
    public interface IPipeReader : IDisposable
    {
        // returns default if Pipe is closed (including when requestSize cannot be fulfilled)
        // requestSize = 0 is unbounded read which
        // internally must have a maximum (like 4K) or else it will subvert DOS protection
        // there is only one active read so a read request subsumes the previous
        // that way a client can read a bit and then request a symantic frame
        ValueTask<IPipeReadBuffer> ReadAsync(int requestSize = 0);
        // no longer an active read
        void Release(int consumed);
    }

    public sealed class SomeClass
    { 
        private const int HeaderSize = 24;
        private const int MaxBodySize = 42 * 42;

        private int ProcessHeader(IPipeReadBuffer buffer)
        {
            return default == buffer ? -1 : 42;
        }

        private bool ProcessBody(IPipeReadBuffer buffer)
        {
            return default != buffer;
        }

        private async ValueTask<bool> ReadSomeStuffAsync(IPipeReader reader)
        {
            using (reader)
            {
                int bodySize = ProcessHeader(await reader.ReadAsync(HeaderSize).ConfigureAwait(false));
                return
                    -1 != bodySize
                    && ProcessBody(await reader.ReadAsync(bodySize).ConfigureAwait(false));
            }
        }

        private readonly Memory<byte> Header = new byte[HeaderSize];

        private IEnumerable<(int hint, Func<IPipeWriteBuffer, int> fill)> ChunkedBody()
        {
            var body = new Memory<byte>(new byte[88]);

            while (0 < body.Length)
            {
                var letsWrite = Math.Min(18, body.Length);
                yield return (44, (b) => { body.Slice(0, letsWrite).CopyTo(b.Slice(0, letsWrite)); return letsWrite; });
                body = body.Slice(letsWrite);
            }
        }

        private async ValueTask<bool> WriteSomeStuffAsync(IPipeWriter writer)
        {
            using (writer)
            {
                return
                    await writer.SendAsync(
                        HeaderSize,
                        (b) => { Header.CopyTo(b.AsMemory()); return HeaderSize; })
                        .ConfigureAwait(false)
                    && await writer.SendAsync(ChunkedBody()).ConfigureAwait(false);
            }
        }
    }

**Comment by davidfowl** [comment]
We're going to be making this change as part of 3.0 https://github.com/dotnet/corefx/issues/35703. Closing this issue out. Thanks for the discussion.
